Wikipedia/Code/ArticleViewController+Announcements.swift

```diff
@@ -0,0 +1,45 @@
+import WMF
+
+
+extension ArticleViewController {
+    func showAnnouncementIfNeeded() {
+        guard UserDefaults.standard.shouldCheckForArticleAnnouncements else {
+            return
+        }
+        let predicate = NSPredicate(format: "placement == %@", "article")
+        let contentGroup = dataStore.viewContext.newestVisibleGroup(of: .announcement, with: predicate)
+        guard
+            let contentGroupURL = contentGroup?.url,
+            let announcement = contentGroup?.contentPreview as? WMFAnnouncement,
+            let actionURL = announcement.actionURL
+        else {
+            UserDefaults.standard.shouldCheckForArticleAnnouncements = false
+            return
+        }
+        let dismiss = {
+            UserDefaults.standard.shouldCheckForArticleAnnouncements = false
+            // re-fetch since time has elapsed
+            let contentGroup = self.dataStore.viewContext.contentGroup(for: contentGroupURL)
+            contentGroup?.markDismissed()
+            contentGroup?.updateVisibilityForUserIsLogged(in: self.session.isAuthenticated)
+            do {
+                try self.dataStore.viewContext.save()
+            } catch let saveError {
+                DDLogError("Error saving after marking article announcement as dismissed: \(saveError)")
+            }
+        }
+        let context = FeedFunnelContext(contentGroup)
+        FeedFunnel.shared.logFeedImpression(for: context)
+        wmf_showAnnouncementPanel(announcement: announcement, primaryButtonTapHandler: { (sender) in
+            self.navigate(to: actionURL, useSafari: true)
+            // dismiss handler is called
```

Comment:
@joewalsh in `wmf_showAnnouncementPanel`, `dismissHandler` is only called in the footer link action, so I think you do need to explicitly call `dismiss()` here and in the secondary tap.

Reply:
I noticed when debugging that these were called twice. I verified now that the other blocks call `self.dismiss()` and the `dismissHandler` is called on `viewDidDisappear`. Thanks for looking out for that though. I will add a comment in the other file as well.

Reply:
Added in the next PR - 802a56bb1